<?xml version='1.0' ?>

<launch>

  <arg name="use_sim_time" default="false" description="Use the /clock topic for time to sync with simulation"/>

  <arg name="control_type" description="Fleet adapter type: full_control or read_only"/>
  <arg name="fleet_name" description="Name of the fleet that this adapter will interface with"/>
  <!-- Note: We give a blank default to nav_graph_file because it is not required for read_only
             mode. However it is still required for full_control mode.
       TODO(MXG): Investigate if there is a better way to handle conditionally required arguments. -->
  <arg name="nav_graph_file" default="" description="The file path of this fleet's navigation graph"/>
  <arg name="linear_velocity" description="The nominal linear velocity of the vehicles in this fleet"/>
  <arg name="angular_velocity" description="The nominal angular velocity of the vehicles in this fleet"/>
  <arg name="linear_acceleration" description="The nominal linear acceleration of the vehicles in this fleet"/>
  <arg name="angular_acceleration" description="The nominal angular acceleration of the vehicles in this fleet"/>
  <arg name="footprint_radius" description="The radius of the footprint of the vehicles in this fleet"/>
  <arg name="vicinity_radius" description="The radius of the personal vicinity of the vehicles in this fleet"/>
  <arg name="perform_deliveries" default="false" description="Whether this fleet adapter can perform deliveries"/>
  <arg name="delay_threshold" default="10.0" description="How long to wait before replanning" />
  <arg name="disable_delay_threshold" default="false" description="Disable the delay_threshold behavior" />
  <arg name="retry_wait" default="10.0" description="How long a retry should wait before starting"/>
  <arg name="discovery_timeout" default="10.0" description="How long to wait on discovery before giving up"/>
  <arg name="reversible" default="true" description="Can the robot drive backwards"/>
  <arg name="output" default="screen"/>

  <arg name="perform_loop" default="false" description="Whether this fleet adapter can perform loops"/>
  <arg name="perform_cleaning" default="false" description="Whether this fleet adapter can perform cleaning"/>
  <arg name="battery_voltage" description="The nominal voltage(V) of the battery in the vehicles in this fleet"/>
  <arg name="battery_capacity" description="The nominal capacity(Ah) of the battery in the vehicles in this fleet"/>
  <arg name="battery_charging_current" description="The nominal charging current(A) of the battery in the vehicles in this fleet"/>
  <arg name="mass" description="The mass(kg) of the vehicles in this fleet"/>
  <arg name="inertia" description="The inertia(kg.m^2) of the vehicles in this fleet"/>
  <arg name="friction_coefficient" description="The friction coefficient of the vehicles in this fleet"/>
  <arg name="ambient_power_drain" description="The power rating(W) of ambient devices (processors, sensors, etc.) of the vehicles in this fleet"/>
  <arg name="tool_power_drain" description="The power rating(W) of special tools (vaccuum, cleaning systems, etc.) of the vehicles in this fleet"/>
  <arg name="drain_battery" default="false" description="Whether battery drain should be considered while assigning tasks to vechiles in this fleet"/>
  <arg name="recharge_threshold" default="0.2" description="The fraction of total battery capacity below which the robot must return to its charger"/>
  <arg name="experimental_lift_watchdog_service" default="" description="(Experimental) The name of a service to check whether a robot can enter a lift"/>


  <node pkg="rmf_fleet_adapter"
        exec="$(var control_type)"
        name="$(var fleet_name)_fleet_adapter"
        output="both">

    <param name="fleet_name" value="$(var fleet_name)"/>

    <param name="nav_graph_file" value="$(var nav_graph_file)"/>

    <param name="linear_velocity" value="$(var linear_velocity)"/>
    <param name="angular_velocity" value="$(var angular_velocity)"/>

    <param name="linear_acceleration" value="$(var linear_acceleration)"/>
    <param name="angular_acceleration" value="$(var angular_acceleration)"/>

    <param name="footprint_radius" value="$(var footprint_radius)"/>
    <param name="vicinity_radius" value="$(var vicinity_radius)"/>

    <param name="perform_deliveries" value="$(var perform_deliveries)"/>
    <param name="perform_loop" value="$(var perform_loop)"/>
    <param name="perform_cleaning" value="$(var perform_cleaning)"/>

    <param name="delay_threshold" value="$(var delay_threshold)"/>
    <param name="disable_delay_threshold" value="$(var disable_delay_threshold)"/>
    <param name="retry_wait" value="$(var retry_wait)"/>
    <param name="discovery_timeout" value="$(var discovery_timeout)"/>
    <param name="reversible" value="$(var reversible)"/>

    <param name="battery_voltage" value="$(var battery_voltage)"/>
    <param name="battery_capacity" value="$(var battery_capacity)"/>
    <param name="battery_charging_current" value="$(var battery_charging_current)"/>
    <param name="mass" value="$(var mass)"/>
    <param name="inertia" value="$(var inertia)"/>
    <param name="friction_coefficient" value="$(var friction_coefficient)"/>
    <param name="ambient_power_drain" value="$(var ambient_power_drain)"/>
    <param name="tool_power_drain" value="$(var tool_power_drain)"/>
    <param name="drain_battery" value="$(var drain_battery)"/>
    <param name="recharge_threshold" value="$(var recharge_threshold)"/> 

    <param name="experimental_lift_watchdog_service" value="$(var experimental_lift_watchdog_service)"/>

    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node>

</launch>
